File: python/example/cpp_reference/src/deepseek_moeinfer.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file deepseek_moeinfer.cpp
 * \brief
 */
#include "interface/function/function.h"
#include "interface/configs/config_manager.h"


namespace npu::tile_fwk {

constexpr float F_1 = 1.0;
constexpr float F_NEGA_1 = -1.0;
constexpr int NUM_32 = 32;
constexpr int NUM_64 = 64;
constexpr int NUM_128 = 128;
constexpr int NUM_2048 = 2048;
constexpr int NUM_7168 = 7168;

void DynamicFFN(const Tensor &hiddenStates, const Tensor &ffnWeight1, const Tensor &ffnWeight2,
                const Tensor &ffnWeight3, Tensor &out, int basicBatch)
{
    int h = hiddenStates.GetShape(1);
    if (basicBatch == 0) {
        throw std::invalid_argument("basic_batch is zero");
    }
    FUNCTION("main", {hiddenStates, ffnWeight1, ffnWeight2, ffnWeight3}, {out}) {
        LOOP("L0", FunctionType::DYNAMIC_LOOP, loopIdx,
                    LoopRange(CeilDiv(GetInputShape(hiddenStates, 0), basicBatch))) {
            SymbolicScalar batchIdx = basicBatch * loopIdx;
            auto hiddenStatesTemp = View(hiddenStates, {basicBatch, h}, {batchIdx, 0});
            auto castRes = Cast(hiddenStatesTemp, DataType::DT_FP16);
            auto gate = Matrix::Matmul(DataType::DT_FP32, castRes, ffnWeight1);
            auto swish = MulS(gate, Element(DataType::DT_FP32, F_NEGA_1));
            swish = Exp(swish);
            swish = AddS(swish, Element(DataType::DT_FP32, F_1));
            swish = Div(gate, swish);

            auto up = Matrix::Matmul(DataType::DT_FP32, castRes, ffnWeight2);
            swish = Mul(swish, up);
            auto swishFp16 = Cast(swish, DataType::DT_FP16);

            // down_proj
            auto mlpRes = Matrix::Matmul<false, true>(DataType::DT_FP32, swishFp16, ffnWeight3);
            Assemble(mlpRes, {batchIdx, 0}, out);
        }
    }
}

void DynamicFFNQuant(const Tensor &hiddenStatesQuant, const Tensor &hiddenStatesScale, const Tensor &ffnWeight1,
                     const Tensor &ffnWeight2, const Tensor &ffnWeight3, const Tensor &ffnScale1,
                     const Tensor &ffnScale2, const Tensor &ffnScale3, Tensor &out, int basicBatch)
{
    int h = hiddenStatesQuant.GetShape(1);
    if (basicBatch == 0) {
        throw std::invalid_argument("basic_batch is zero");
    }
    FUNCTION("main", {hiddenStatesQuant, hiddenStatesScale, ffnWeight1, ffnWeight2, ffnWeight3, ffnScale1,
             ffnScale2, ffnScale3}, {out}) {
        LOOP("L0", FunctionType::DYNAMIC_LOOP, loopIdx,
                    LoopRange(CeilDiv(GetInputShape(hiddenStatesQuant, 0), basicBatch))) {
            SymbolicScalar batchIdx = basicBatch * loopIdx;

            auto castRes = View(hiddenStatesQuant, {basicBatch, h}, {batchIdx, 0});
            auto castResScale = View(hiddenStatesScale, {basicBatch, 1}, {batchIdx, 0});
            auto gateInt32 = Matrix::Matmul(DataType::DT_INT32, castRes, ffnWeight1);

            // dequant: int32 -> fp32 -> *scale -> fp16/bf16
            auto gateTmpFp32 = Cast(gateInt32, DataType::DT_FP32);
            auto gateTmpDequantPerToken = Mul(gateTmpFp32, castResScale);
            auto gate = Mul(gateTmpDequantPerToken, ffnScale1);

            // swish: x / (1 + e^(-x))
            auto swish = MulS(gate, Element(DataType::DT_FP32, F_NEGA_1));
            swish = Exp(swish);
            swish = AddS(swish, Element(DataType::DT_FP32, F_1));
            swish = Div(gate, swish);

            auto upInt32 = Matrix::Matmul(DataType::DT_INT32, castRes, ffnWeight2);
            // upProj
            auto upTmpFp32 = Cast(upInt32, DataType::DT_FP32);
            auto upTmpDequantPerToken = Mul(upTmpFp32, castResScale);
            auto up = Mul(upTmpDequantPerToken, ffnScale2);

            swish = Mul(swish, up);

            // downProj
            auto swishQuantRes = Quant(swish); // int8
            Tensor swishRes = std::get<0>(swishQuantRes);
            Tensor swishScale = std::get<1>(swishQuantRes);

            Tensor resInt32 = Matrix::Matmul<false, true>(DataType::DT_INT32, swishRes, ffnWeight3);
            auto resTmpFp32 = Cast(resInt32, DataType::DT_FP32);
            auto resTmpDequantPerToken = Mul(resTmpFp32, swishScale);
            auto res = Mul(resTmpDequantPerToken, ffnScale3);
            Assemble(res, {batchIdx, 0}, out);
        }
    }
}

void TestDynamicFFN()
{
    config::SetHostOption(ONLY_CODEGEN, true);

    TileShape::Current().SetVecTile(NUM_32, NUM_128);
    TileShape::Current().SetCubeTile({NUM_32, NUM_32}, {NUM_128, NUM_128}, {NUM_128, NUM_128});

    constexpr int batchSize = NUM_64;
    constexpr int sequence = 1;
    constexpr int h = NUM_7168;
    constexpr int expertDim = NUM_2048;
    constexpr int bs = batchSize * sequence;
    constexpr int basicBatch = NUM_32;

    std::vector<int64_t> hiddenStatesShape{bs, h};
    std::vector<int64_t> weightShape{h, expertDim};
    std::vector<int64_t> outShape{bs, h};

    Tensor hiddenStates(DT_FP32, hiddenStatesShape, "hiddenStates");
    Tensor ffnWeight1(DT_FP16, weightShape, "weightShape1");
    Tensor ffnWeight2(DT_FP16, weightShape, "weightShape2");
    Tensor ffnWeight3(DT_FP16, weightShape, "weightShape3");
    Tensor ffnOut(DT_FP32, outShape, "ffnout");

    DynamicFFN(hiddenStates, ffnWeight1, ffnWeight2, ffnWeight3, ffnOut, basicBatch);
}

void TestDynamicFFNQuant()
{
    config::SetHostOption(ONLY_CODEGEN, true);
    TileShape::Current().SetVecTile(NUM_32, NUM_128);
    TileShape::Current().SetCubeTile({NUM_32, NUM_32}, {NUM_128, NUM_128}, {NUM_128, NUM_128});

    constexpr int batchSize = NUM_32;
    constexpr int sequence = 1;
    constexpr int h = NUM_7168;
    constexpr int expertDim = NUM_2048;
    constexpr int bs = batchSize * sequence;
    constexpr int basicBatch = NUM_32;

    std::vector<int64_t> hiddenStatesShape{bs, h};
    std::vector<int64_t> weightShape{h, expertDim};
    std::vector<int64_t> outShape{bs, h};

    Tensor hiddenStates(DT_INT8, hiddenStatesShape, "hiddenStates");
    Tensor hiddenStatesScale(DT_FP32, {bs, 1}, "hiddenStatesScale");
    Tensor ffnWeight1(DT_INT8, weightShape, "ffnWeight1", TileOpFormat::TILEOP_NZ);
    Tensor ffnWeight2(DT_INT8, weightShape, "ffnWeight2", TileOpFormat::TILEOP_NZ);
    Tensor ffnWeight3(DT_INT8, weightShape, "ffnWeight3", TileOpFormat::TILEOP_NZ);
    Tensor ffnScale1(DT_FP32, {1, expertDim}, "ffnScale1");
    Tensor ffnScale2(DT_FP32, {1, expertDim}, "ffnScale2");
    Tensor ffnScale3(DT_FP32, {1, h}, "ffnScale3");
    Tensor ffnOut(DT_FP32, outShape, "ffnout");

    DynamicFFNQuant(hiddenStates, hiddenStatesScale, ffnWeight1, ffnWeight2, ffnWeight3, ffnScale1, ffnScale2,
        ffnScale3, ffnOut, basicBatch);
}
} // namespace npu::tile_fwk

File: python/example/deepseek_moeinfer.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ======================================================================================================================
"""
"""
import pto

# RuntimeConfig KEYS
KEY_ONLY_CODEGEN = "only_codegen"

F_1 = 1.0
F_NEGA_1 = -1.0


def ceil_div(a, b):
    return (a + (b - 1)) // b


def dynamic_ffn(**kwargs):
    hidden_states = kwargs.get("hidden_states")
    ffn_weight1 = kwargs.get("ffn_weight1")
    ffn_weight2 = kwargs.get("ffn_weight2")
    ffn_weight3 = kwargs.get("ffn_weight3")
    out = kwargs.get("out")
    basic_batch = kwargs.get("basic_batch")

    h = hidden_states.shape[1]

    input_tensors = [hidden_states, ffn_weight1, ffn_weight2, ffn_weight3]
    output_tensors = [out]
    if basic_batch == 0:
        raise ValueError("basic_batch is zero")
    with pto.function("main", input_tensors, output_tensors):
        for idx in pto.loop(ceil_div(pto.get_input_shape(hidden_states, 0), basic_batch), name="L0", idx_name="idx"):
            def inside_idx_loop(idx):
                batch_idx = basic_batch * idx
                hidden_states_temp = pto.view(hidden_states, [basic_batch, h], [batch_idx, 0])
                cast_res = pto.cast(hidden_states_temp, pto.DT_FP16)
                gate = pto.matmul(cast_res, ffn_weight1, pto.DT_FP32)
                swish = pto.mul_s(gate, pto.element(pto.DT_FP32, F_NEGA_1))
                swish = pto.exp(swish)
                swish = pto.add_s(swish, pto.element(pto.DT_FP32, F_1))
                swish = pto.div(gate, swish)

                up = pto.matmul(cast_res, ffn_weight2, pto.DT_FP32)
                swish = pto.mul(swish, up)
                swish_fp16 = pto.cast(swish, pto.DT_FP16)

                # down_proj
                mlp_res = pto.matmul(swish_fp16, ffn_weight3, pto.DT_FP32, a_trans=False, b_trans=True)
                pto.assemble(mlp_res, [batch_idx, 0], out)
            inside_idx_loop(idx)


def dynamic_ffn_quant(**kwargs):
    hidden_states_quant = kwargs.get("hidden_states_quant")
    hidden_states_scale = kwargs.get("hidden_states_scale")
    ffn_weight1 = kwargs.get("ffn_weight1")
    ffn_weight2 = kwargs.get("ffn_weight2")
    ffn_weight3 = kwargs.get("ffn_weight3")
    ffn_scale1 = kwargs.get("ffn_scale1")
    ffn_scale2 = kwargs.get("ffn_scale2")
    ffn_scale3 = kwargs.get("ffn_scale3")
    out = kwargs.get("out")
    basic_batch = kwargs.get("basic_batch")

    h = hidden_states_quant.shape[1]

    input_tensors = [hidden_states_quant, hidden_states_scale, ffn_weight1, ffn_weight2,
                    ffn_weight3, ffn_scale1, ffn_scale2, ffn_scale3]
    output_tensors = [out]
    if basic_batch == 0:
        raise ValueError("basic_batch is zero")
    with pto.function("main", input_tensors, output_tensors):
        for idx in pto.loop(ceil_div(pto.get_input_shape(hidden_states_quant, 0), basic_batch),
        name="L0", idx_name="idx"):
            def inside_idx_loop(idx):
                batch_idx = basic_batch * idx

                cast_res = pto.view(hidden_states_quant, [basic_batch, h], [batch_idx, 0])
                cast_res_scale = pto.view(hidden_states_scale, [basic_batch, 1], [batch_idx, 0])
                gate_int32 = pto.matmul(cast_res, ffn_weight1, pto.DT_INT32)

                # dequant: int32 -> fp32 -> *scale -> fp16/bf16
                gate_tmp_fp32 = pto.cast(gate_int32, pto.DT_FP32)
                gate_tmp_dequant_pertoken = pto.mul(gate_tmp_fp32, cast_res_scale)
                gate = pto.mul(gate_tmp_dequant_pertoken, ffn_scale1)

                swish = pto.mul_s(gate, pto.element(pto.DT_FP32, F_NEGA_1))
                swish = pto.exp(swish)
                swish = pto.add_s(swish, pto.element(pto.DT_FP32, F_1))
                swish = pto.div(gate, swish)

                up_int32 = pto.matmul(cast_res, ffn_weight2, pto.DT_INT32)
                # upProj
                up_tmp_fp32 = pto.cast(up_int32, pto.DT_FP32)
                up_tmp_dequant_pertoken = pto.mul(up_tmp_fp32, cast_res_scale)
                up = pto.mul(up_tmp_dequant_pertoken, ffn_scale2)

                swish = pto.mul(swish, up)

                # downProj
                swish_quant_res = pto.quant(swish) # int8
                swish_res = swish_quant_res[0]
                swish_scale = swish_quant_res[1]

                res_int32 = pto.matmul(swish_res, ffn_weight3, pto.DT_INT32, a_trans=False, b_trans=True)
                res_tmp_fp32 = pto.cast(res_int32, pto.DT_FP32)
                res_tmp_dequant_pertoken = pto.mul(res_tmp_fp32, swish_scale)
                res = pto.mul(res_tmp_dequant_pertoken, ffn_scale3)
                pto.assemble(res, [batch_idx, 0], out)
            inside_idx_loop(idx)


def test_ffn():
    pto.set_host_option(KEY_ONLY_CODEGEN, True)

    pto.set_vec_tile_shapes(32, 128)
    pto.set_cube_tile_shapes([32, 32], [128, 128], [128, 128])

    batch_size = 64
    sequence = 1
    h = 7168
    expert_dim = 2048
    bs = batch_size * sequence
    basic_batch = 32

    hidden_states_shape = [bs, h]
    weight_shape = [h, expert_dim]
    out_shape = [bs, h]

    hidden_states = pto.tensor(hidden_states_shape, pto.DT_FP32, "hidden_states")
    ffn_weight1 = pto.tensor(weight_shape, pto.DT_FP16, "weight_shape1")
    ffn_weight2 = pto.tensor(weight_shape, pto.DT_FP16, "weight_shape2")
    ffn_weight3 = pto.tensor(weight_shape, pto.DT_FP16, "weight_shape3")
    ffn_out = pto.tensor(out_shape, pto.DT_FP32, "ffn_out")

    dynamic_ffn(hidden_states=hidden_states,
                ffn_weight1=ffn_weight1,
                ffn_weight2=ffn_weight2,
                ffn_weight3=ffn_weight3,
                out=ffn_out,
                basic_batch=basic_batch)


def test_ffn_quant():
    pto.set_host_option(KEY_ONLY_CODEGEN, True)

    pto.set_vec_tile_shapes(32, 128)
    pto.set_cube_tile_shapes([32, 32], [128, 128], [128, 128])

    batch_size = 32
    sequence = 1
    h = 7168
    expert_dim = 2048
    bs = batch_size * sequence
    basic_batch = 32

    hidden_states_shape = [bs, h]
    weight_shape = [h, expert_dim]
    out_shape = [bs, h]

    hidden_states_quant = pto.tensor(hidden_states_shape, pto.DT_INT8, "hidden_states_quant")
    hidden_states_scale = pto.tensor([bs, 1], pto.DT_FP32, "hidden_states_scale")
    ffn_weight1 = pto.tensor(weight_shape, pto.DT_INT8, "weight_shape1", pto.TILEOP_NZ)
    ffn_weight2 = pto.tensor(weight_shape, pto.DT_INT8, "weight_shape2", pto.TILEOP_NZ)
    ffn_weight3 = pto.tensor(weight_shape, pto.DT_INT8, "weight_shape3", pto.TILEOP_NZ)
    ffn_scale1 = pto.tensor([1, expert_dim], pto.DT_FP32, "ffn_scale1")
    ffn_scale2 = pto.tensor([1, expert_dim], pto.DT_FP32, "ffn_scale2")
    ffn_scale3 = pto.tensor([1, h], pto.DT_FP32, "ffn_scale3")
    ffn_out = pto.tensor(out_shape, pto.DT_FP32, "ffn_out")

    dynamic_ffn_quant(hidden_states_quant=hidden_states_quant,
                    hidden_states_scale=hidden_states_scale,
                    ffn_weight1=ffn_weight1,
                    ffn_weight2=ffn_weight2,
                    ffn_weight3=ffn_weight3,
                    ffn_scale1=ffn_scale1,
                    ffn_scale2=ffn_scale2,
                    ffn_scale3=ffn_scale3,
                    out=ffn_out,
                    basic_batch=basic_batch)